SpatialGDK/Source/SpatialGDK/Public/LoadBalancing/ReferenceCountedLockingPolicy.h

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "AbstractLockingPolicy.h"
+#include "Containers/Map.h"
+#include "Containers/UnrealString.h"
+#include "GameFramework/Actor.h"
+
+#include "ReferenceCountedLockingPolicy.generated.h"
+
+DECLARE_LOG_CATEGORY_EXTERN(LogReferenceCountedLockingPolicy, Log, All)
+
+UCLASS()
+class SPATIALGDK_API UReferenceCountedLockingPolicy : public UAbstractLockingPolicy
+{
+	GENERATED_BODY()
+
+public:
+	virtual bool CanAcquireLock(const AActor* Actor) const override;
+
+	virtual ActorLockToken AcquireLock(const AActor* Actor, FString DebugString = "") override;
+
+	// This should only be called during the lifetime of the locked actor
+	virtual void ReleaseLock(ActorLockToken Token) override;
+
+	virtual bool IsLocked(const AActor* Actor) const override;
+
+private:
+	struct LockNameAndActor
+	{
+		FString LockName;
+		const AActor* Actor;
+	};
+
+	TMap<const AActor*, int32> ActorToReferenceCount;
```

Comment:
What happens if an Actor is deleted? Don't we need to clean up these structures?

Reply:
It's really hard for us to detect this at the moment. Currently it's the responsibility of whoever is locking to not delete the actor when they use it. I meant to look into better ways of trying to detect this though.

Reply:
Actually I may have lied / missed something obvious.

Reply:
Ok we can detect stuff provided users don't override a function.